### PR TITLE
Python: Fix ChatHistoryTruncationReducer deleting system prompt

### LINKED ISSUE #12612: Python: Bug: Python 1.34.0 ChatHistoryTruncationReducer deletes the System Prompt
**Describe the bug**
When using the `ChatHistoryTruncationReducer` the reducer reduces the system prompt (both if added to the constructor via `system_message` or when manually added as first message to the history )

**To Reproduce**
1. Create a History using the ChatHistoryTruncationReducer, and add a `target_count`
2. populate the ChatHistory with more elements than the `target_count`
3. call the reduce function

**Expected behavior**
I expect the System Prompt to be kept, as it is an integral part of the conversation

**Screenshots**
see first comment.

**Platform**
 - Language: Python
 - Source: pip semantic-kernel 1.34.0
 - AI model:  OpenAI:GPT-4o-mini(2024-12-01-preview)
 - IDE:  VS Code
 - OS:  Windows

**Additional context**
Nothing, but additional context can be provided if needed.

### PR DESCRIPTION
**Pull request by roli-lpci**
## Summary

`ChatHistoryTruncationReducer.reduce()` silently deletes system/developer messages when truncating chat history. This is because it calls `extract_range()`, which unconditionally filters out system/developer messages — a function designed for the summarization use case, not truncation.

Fixes #12612.

## Approach

Port the .NET SDK fix (PR #10344) to Python:

1. Detect the first system/developer message before truncation
2. Pass `has_system_message=True` to `locate_safe_reduction_index` so `target_count` accounts for the preserved message (matches .NET's `targetCount -= hasSystemMessage ? 1 : 0`)
3. Use a simple `history[truncation_index:]` slice instead of `extract_range` (which strips system messages)
4. Prepend the system message if it was truncated away

Also adds a guard for `target_count <= 0` after the system message adjustment to prevent `IndexError` when `target_count=1`.

## Changes

| File | Change |
|------|--------|
| `chat_history_reducer_utils.py` | Add `has_system_message` parameter, adjust `target_count`, guard against `<= 0` |
| `chat_history_truncation_reducer.py` | Detect system message, bypass `extract_range`, prepend if truncated |
| `test_chat_history_truncation_reducer.py` | Update 2 existing tests + 4 new tests (system, developer, no-system, target_count=1) |

## Note

The summarization reducer (`ChatHistorySummarizationReducer`) has the same bug — it also uses `extract_range` and loses system messages during summarization. The .NET summarization reducer preserves system messages via `AssemblySummarizedHistory`. This should be addressed in a follow-up PR to keep this change focused.

## Test plan

- [x] 11/11 truncation reducer tests pass (was 7, now 11)
- [x] 9/9 utils tests pass (no regressions from new parameter)
- [x] 11/11 summarization reducer tests pass (unaffected)
- [x] Reproduces and fixes the exact scenario from #12612
- [x] Edge case: `target_count=1` with system message does not crash

### CODE REVIEW

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 3 | **Confidence:** 85%

### ✗ Correctness
> The diff correctly preserves system/developer messages during truncation by adjusting `target_count` and re-prepending the system message when it gets sliced away. The approach is sound for the common case. However, there is a correctness issue in the edge case where `target_count=1` with a system message: `locate_safe_reduction_index` returns `None` (meaning 'no reduction needed'), so the history is never reduced even when it contains many messages. The test for this case is lenient and accepts `None`, masking the bug. Additionally, the `ChatHistorySummarizationReducer` is not updated with the same system-message preservation logic, which may lead to inconsistent behavior between the two reducers.

### ✓ Security Reliability
> The diff preserves system/developer messages during chat history truncation, aligning with .NET SDK behavior. The changes are generally sound from a security perspective — no injection risks, unsafe deserialization, or secrets. The main reliability concerns are: (1) only the first system/developer message is preserved, silently dropping any subsequent ones; (2) when target_count=1 with a system message, reduction is silently disabled forever, allowing unbounded history growth; and (3) the identity-based membership check for re-prepending the system message could behave unexpectedly if message equality is value-based rather than identity-based.

### ✗ Test Coverage
> The diff adds system/developer message preservation to the truncation reducer and includes several new tests. Overall coverage is good, but there is one test (`test_truncation_target_count_1_with_system_message`) whose key assertion is unreachable dead code — the `if result is not None` guard means the test always passes vacuously. Additionally, there are no direct unit tests for the modified `locate_safe_reduction_index` utility with the new `has_system_message` parameter, and no tests cover edge cases like multiple system/developer messages or the `ChatHistorySummarizationReducer` which also calls the utility.

### Blocking Issues
- [ ] When `target_count=1` and `has_system_message=True`, `locate_safe_reduction_index` returns `None` (no reduction), leaving the entire history intact even when it far exceeds the target. The caller should still be able to reduce to just the system message in this scenario. The early `return None` when `target_count <= 0` prevents a crash (the scan loop would IndexError at `history[total_count]`), but the behavior is wrong — the reducer should produce `[system_message]` with 1 total message, not silently skip reduction.
- [ ] test_truncation_target_count_1_with_system_message (lines 161-170): The assertion inside `if result is not None` is dead code. With target_count=1 and a system message, `locate_safe_reduction_index` decrements target_count to 0 and returns None, so `reduce()` always returns None. The test gives a false sense of coverage. It should explicitly assert `result is None` and, ideally, add a separate case that actually exercises the preservation path.

### Suggestions
- The `ChatHistorySummarizationReducer` also calls `locate_safe_reduction_index` but is not updated to pass `has_system_message` or to preserve system messages. Consider applying the same fix there for consistency.
- The check `system_message not in truncated_list` uses value equality (`__eq__`). If `ChatMessageContent` has deep equality semantics (e.g., Pydantic models), a different message with identical fields could cause a false match. Consider using identity comparison (`all(msg is not system_message for msg in truncated_list)`) for robustness.
- The test `test_truncation_target_count_1_with_system_message` is overly lenient — it accepts `None` (no reduction) as valid, which masks the bug. The test should assert that reduction occurs and the result contains only the system message.
- Only the first system/developer message is found via `next(...)`. If a history legitimately contains multiple system or developer messages, all but the first will be silently dropped on truncation. Consider collecting all such messages or documenting this as a known limitation.
- When `target_count=1` and a system message exists, `locate_safe_reduction_index` returns `None` unconditionally, meaning truncation never fires and history grows unboundedly. Consider logging a warning so users can detect this misconfiguration.
- The check `system_message not in truncated_list` relies on the equality semantics of message objects. If `ChatMessageContent.__eq__` is value-based, a different message with identical content could prevent the system message from being re-prepended. Using identity comparison (`all(msg is not system_message for msg in truncated_list)`) would be more robust.
- Add direct unit tests for `locate_safe_reduction_index` with `has_system_message=True`, covering the target_count adjustment and the target_count<=0 early return.
- Add a test with multiple system/developer messages to verify only the first is preserved (current `next()` semantics).
- Add a test where the system message falls within the retained tail (i.e., system message is NOT truncated away) to exercise the `system_message not in truncated_list` false branch.
- Verify whether `ChatHistorySummarizationReducer` also needs updating to pass `has_system_message` to `locate_safe_reduction_index`, and add corresponding tests if so.

---
*Automated review by moonbox3's agents*

File: python/semantic_kernel/contents/history_reducer/chat_history_reducer_utils.py [hunk 1]
```diff
@@ -83,11 +84,21 @@ def locate_safe_reduction_index(
...
         offset_count: Optional number of messages to skip at the start (e.g. existing summary messages).
+        has_system_message: Whether the history contains a system message that will be preserved
+                           separately. When True, the target_count is adjusted to account for the
+                           system message being re-added after reduction.
 
     Returns:
         The index that identifies the starting point for a reduced history that does not orphan
         sensitive content. Returns None if reduction is not needed.
     """
+    # Adjust target_count to account for the system message that will be preserved separately.
```
Comment (moonbox3) [inline python/semantic_kernel/contents/history_reducer/chat_history_reducer_utils.py:95]: When `has_system_message=True` and `target_count` was 1, this decrements to 0 and returns `None` (no reduction). But the history may have many messages that should be reduced — the reducer should produce a result containing only the system message. Returning `None` here causes the truncation reducer to silently skip reduction entirely. Consider handling `target_count == 0` as 'truncate everything' (return `len(history)`) and let the caller prepend the system message, or handle this edge case in the truncation reducer itself.

File: python/semantic_kernel/contents/history_reducer/chat_history_truncation_reducer.py [hunk 2]
```diff
@@ -45,15 +45,32 @@ async def reduce(self) -> Self | None:
...
                 f"No truncation index found. Target count: {self.target_count}, Threshold: {self.threshold_count}"
             )
             return None
 
         logger.info(f"Truncating history to {truncation_index} messages.")
-        truncated_list = extract_range(history, start=truncation_index)
+        truncated_list = history[truncation_index:]
+
+        # Prepend the system/developer message if it was truncated away
+        if system_message is not None and system_message not in truncated_list:
```
Comment (moonbox3) [inline python/semantic_kernel/contents/history_reducer/chat_history_truncation_reducer.py]: This uses value equality (`__eq__`) via `not in`. Since `system_message` is a reference into the original `history` list and `truncated_list` is a slice of the same list, identity comparison would be more precise and avoid a potential false match if another message happens to have identical fields.

```suggestion
        if system_message is not None and all(msg is not system_message for msg in truncated_list):
```

File: python/tests/unit/contents/test_chat_history_truncation_reducer.py [hunk 3]
```diff
@@ -96,12 +98,89 @@ async def test_truncation_reducer_truncation_with_tools(chat_messages_with_tools
...
+async def test_truncation_target_count_1_with_system_message():
+    """Verify target_count=1 with system message does not crash (edge case from review)."""
+    reducer = ChatHistoryTruncationReducer(target_count=1, system_message="System prompt")
+    reducer.add_message(ChatMessageContent(role=AuthorRole.USER, content="Hello"))
+    reducer.add_message(ChatMessageContent(role=AuthorRole.ASSISTANT, content="Hi"))
+    reducer.add_message(ChatMessageContent(role=AuthorRole.USER, content="How are you?"))
+    reducer.add_message(ChatMessageContent(role=AuthorRole.ASSISTANT, content="Good"))
+
+    # Should not crash. Either returns None (no reduction possible)
+    # or returns just the system message.
```
Comment (moonbox3) [inline python/tests/unit/contents/test_chat_history_truncation_reducer.py]: This test accepts `None` (no reduction) as a valid outcome for `target_count=1` with a system message, but that behavior is a bug — the user asked for at most 1 message and the history has 5. The test should assert that reduction happens and the result contains exactly the system message. As written, this test will pass even when the reducer fails to reduce.

File: python/semantic_kernel/contents/history_reducer/chat_history_reducer_utils.py (hunk 1 above)
Comment (moonbox3) [inline python/semantic_kernel/contents/history_reducer/chat_history_reducer_utils.py:95]: When `target_count` is 1 and `has_system_message` is True, this permanently prevents any reduction. Consider emitting a warning log so callers can detect this silent no-op, since an ever-growing history could eventually cause OOM or token-limit failures.

```suggestion
        target_count -= 1
        if target_count <= 0:
            import logging
            logging.getLogger(__name__).warning(
                "target_count after accounting for system message is %d; reduction disabled.", target_count
            )
            return None  # Cannot reduce further; only system message would remain
```

File: python/semantic_kernel/contents/history_reducer/chat_history_truncation_reducer.py [hunk 4]
```diff
@@ -45,15 +45,32 @@ async def reduce(self) -> Self | None:
...
             logger.info(
                 f"No truncation index found. Target count: {self.target_count}, Threshold: {self.threshold_count}"
             )
             return None
 
         logger.info(f"Truncating history to {truncation_index} messages.")
-        truncated_list = extract_range(history, start=truncation_index)
+        truncated_list = history[truncation_index:]
+
+        # Prepend the system/developer message if it was truncated away
```
Comment (moonbox3) [inline python/semantic_kernel/contents/history_reducer/chat_history_truncation_reducer.py]: Using `not in` performs value-based equality checks. If `ChatMessageContent.__eq__` compares by content, a duplicate message elsewhere in the list could shadow the original system message and prevent it from being re-prepended. An identity check would be safer here.

```suggestion
        if system_message is not None and not any(m is system_message for m in truncated_list):
```

File: python/tests/unit/contents/test_chat_history_truncation_reducer.py [hunk 5]
```diff
@@ -96,12 +98,89 @@ async def test_truncation_reducer_truncation_with_tools(chat_messages_with_tools
...
+    result = await reducer.reduce()
+    if result is not None:
+        # System message must be preserved
+        assert any(m.role == AuthorRole.SYSTEM for m in result.messages)
+
```
Comment (moonbox3) [inline python/tests/unit/contents/test_chat_history_truncation_reducer.py:174]: Dead assertion: with `target_count=1` and a system message, `locate_safe_reduction_index` reduces target to 0 and returns `None`, so `reduce()` returns `None`. The `if result is not None` branch never executes, meaning the system-message assertion is never checked. Assert the expected behavior explicitly.

```suggestion
    # target_count=1 with system message means adjusted target becomes 0,
    # so no reduction is possible.
    result = await reducer.reduce()
    assert result is None
```

**Comment by roli-lpci** [comment]
Thanks for the thorough review — all three blocking issues were valid. Pushed fixes:

**1. `target_count=1` with system message (unbounded growth)**

`locate_safe_reduction_index` now returns `len(history)` instead of `None` when `target_count <= 0` after the system message adjustment. This produces an empty tail (`history[len(history):]` = `[]`), and the caller prepends the system message, resulting in `[system_message]`. The reducer now correctly reduces to just the system message instead of silently skipping reduction.

Note: this actually improves on the .NET SDK's handling of the same edge case — the .NET implementation has a latent issue where `target_count=1` with a system message can produce unexpected behavior.

**2. Dead test assertion**

Replaced the vacuous `if result is not None` guard with explicit assertions:
```python
assert result is not None
assert len(result.messages) == 1
assert result.messages[0].role == AuthorRole.SYSTEM
assert result.messages[0].content == "System prompt"
```

**3. Identity comparison**

Switched from `system_message not in truncated_list` (value equality) to `all(msg is not system_message for msg in truncated_list)` (identity) to avoid false matches from Pydantic's value-based `__eq__`.

---

Regarding the non-blocking suggestions:

- **ChatHistorySummarizationReducer consistency**: Agreed this should be updated too. Happy to add it in this PR or as a follow-up — let me know your preference.
- **Multiple system/developer messages**: The current `next()` semantics preserve only the first. This matches the .NET SDK behavior. Worth documenting as a known limitation.
- **Direct unit tests for `locate_safe_reduction_index`**: Can add these if desired.

**Comment by moonbox3** [comment]
@roli-lpci please resolve or respond to comments if they are no longer valid, it helps us during our reviews. Thanks!

**Comment by roli-lpci** [comment]
Thanks for the detailed review @moonbox3 — pushed fixes for all remaining comments.

**1. Warning log when reduction disabled (comment on `chat_history_reducer_utils.py:95`)**

Added `logger.warning()` when `target_count <= 0` after system message adjustment, so callers can detect this configuration:
```python
logger.warning(
    "target_count after accounting for system message is %d; reduction will keep only the system message.",
    target_count,
)
```

**2. ChatHistorySummarizationReducer consistency**

Extended the same system/developer message preservation to the summarization reducer. Key detail: summary messages with `SYSTEM` role are excluded via `SUMMARY_METADATA_KEY` filter — they're generated content, not original system prompts.

**3. Double-counting fix (found during internal review)**

Discovered a subtle bug: when the system message falls within the retained tail (e.g., system at index 3, target_count=3, 5 messages), the target_count adjustment produced `target_count - 1` messages instead of `target_count`. The system message was counted twice — once in the -1 adjustment, once as a tail slot.

Fix: `system_would_be_truncated = system_message_index < len(history) - self.target_count` — only adjusts when the system message would actually be lost.

**4. New tests (11 total)**

- 4 direct unit tests for `locate_safe_reduction_index` with `has_system_message`
- Multiple system/developer messages (verifies only first preserved)
- System message in retained tail (verifies no duplication, correct count)
- Warning log verification
- Summarization: SYSTEM preservation, DEVELOPER preservation, SUMMARY_METADATA_KEY exclusion

All 32 tests pass, ruff clean.

**Comment by moonbox3** [comment]
@roli-lpci please resolve comments if they've been addressed, thank you.
